adv-rumors.lua

```diff
@@ -1,83 +1,126 @@
+    -- Clear the string
+    for i, data in ipairs(choice.print_string.text) do
+        df.delete(data)
+    end
+    choice.print_string.text:resize(0)
+
+    -- Split the string assuming \n is newline
+    local line_table = string.split(text, "\n")
+    for i, line in ipairs(line_table) do
+        -- Create a df string for each line
+        local line_ptr = df.new('string')
+        line_ptr.value = line
+        -- Insert it into the text data
+        choice.print_string.text:insert('#', line_ptr)
     end
 end
 
+-- Gets the keywords already present on the dialog choice
+function getKeywords(choice)
+    local keywords = {}
+    for i, keyword in ipairs(choice.key_word) do
+        table.insert(keywords, keyword.value:lower())
+    end
+    return keywords
+end
+
+-- Adds a keyword to the dialog choice
 function addKeyword(choice, keyword)
     local keyword_ptr = df.new('string')
     keyword_ptr.value = keyword
-    choice.keywords:insert('#', keyword_ptr)
+    choice.key_word:insert('#', keyword_ptr)
 end
 
-function rumorUpdate()
-    local improveReadability = true
-    local addKeywordSlew = true
-    local shortenString = true
-    local addKeywordNames = true
-
-    for i, choice in ipairs(df.global.adventure.conversation.choices) do
-        if choice.choice.type == df.talk_choice_type.SummarizeConflict then
-            if improveReadability then
-                condenseChoiceTitle(choice)
-            end
-            if shortenString then
-                condenseChoiceTitle(choice)
-                choice.title[0].value = choice.title[0].value
-                    :gsub("Summarize the conflict in which +", "")
-                    :gsub("This occurred +", "")
-            end
-            if addKeywordSlew then
-                if string.find(choice.title[0].value, "slew") then
-                    addKeyword(choice, 'slew')
-                end
-            end
-            if addKeywordNames then
-                local title = choice.title[0].value
-                for keyword in title:sub(1, title:find('%(') - 1):gmatch('%w+') do
-                    keyword = dfhack.utf2df(dfhack.df2utf(keyword):lower())
-                    if not names_blacklist[keyword] then
-                        addKeyword(choice, keyword)
-                    end
-                end
+-- Adds multiple keywords to the dialog choice
+function addKeywords(choice, keywords)
+    for i, keyword in ipairs(keywords) do
+        addKeyword(choice, keyword)
+    end
+end
+
+-- Generates keywords based on the text of the dialog choice, plus keywords for special cases
+function generateKeywordsForChoice(choice)
+    local new_keywords, keywords_set = {}, utils.invert(getKeywords(choice))
+    for i, data in ipairs(choice.print_string.text) do
+        local text = dfhack.toSearchNormalized(data.value)
+        -- Special cases
+        if not keywords_set['kill'] and string.find(text, "slew") or string.find(text, "slain") then
```

Comment:
this will result in additional "kill" and "slay" keywords being added for every line that has "slain" in it. we might be better off doing these checks when we go through word by word to ensure we match what we think we're matching. (e.g. the above conditional will also match "Akbom Rekoslain".) It might also help simplify the code that checks for duplicates.

For example, the function could look like this:
```lua
-- Generates keywords based on the text of the dialog choice, plus keywords for special cases
function generateKeywordsForChoice(choice)
    local new_keywords, keywords_set = {}, utils.invert(getKeywords(choice))

    local function collect_keyword(word)
        if ignore_words[word] or keywords_set[word] then return end
        table.insert(new_keywords, word)
        keywords_set[word] = true
    end

    -- generate keywords from useful words in the text
    for _, data in ipairs(choice.print_string.text) do
        for word in dfhack.toSearchNormalized(data.value):gmatch('%w+') do
            if word == 'slew' or word == 'slain' then
                collect_keyword('kill')
                collect_keyword('slay')
            elseif word == 'you' or word == 'your' then
                collect_keyword('me')
            else
                collect_keyword(word)
            end
        end
    end
end
```